source/core_mqtt_serializer.c

```diff
@@ -2677,4 +4121,382 @@ MQTTStatus_t MQTT_ProcessIncomingPacketTypeAndLength( const uint8_t * pBuffer,
+            status = MQTT_GetConnectPropertiesSize( pConnectProperties );
+        }
+
+        if( status == MQTTSuccess )
+        {
+            connectPacketSize += pConnectProperties->propertyLength;
+            connectPacketSize += remainingLengthEncodedSize( pConnectProperties->propertyLength );
+            /* Add the length of the client identifier. */
+            connectPacketSize += pConnectInfo->clientIdentifierLength + sizeof( uint16_t );
+
+            /* Add the lengths of the will properties if provided. */
+            if( pWillInfo != NULL )
+            {
+                status = MQTT_GetPublishPropertiesSize( pWillInfo );
+            }
+        }
+
+        if( status == MQTTSuccess )
+        {
+            /* Add the lengths of the will message and topic name if provided. */
+            if( pWillInfo != NULL )
+            {
+                connectPacketSize += pWillInfo->propertyLength;
+                connectPacketSize += remainingLengthEncodedSize( pWillInfo->propertyLength );
+                connectPacketSize += pWillInfo->topicNameLength + sizeof( uint16_t ) +
+                                     pWillInfo->payloadLength + sizeof( uint16_t );
+            }
+
+            /* Add the lengths of the user name and password if provided. */
+            if( pConnectInfo->pUserName != NULL )
+            {
+                connectPacketSize += pConnectInfo->userNameLength + sizeof( uint16_t );
+            }
+
+            if( pConnectInfo->pPassword != NULL )
+            {
+                connectPacketSize += pConnectInfo->passwordLength + sizeof( uint16_t );
+            }
+
+            /* At this point, the "Remaining Length" field of the MQTT CONNECT packet has
+             * been calculated. */
+            remainingLength = connectPacketSize;
+
+            /* Calculate the full size of the MQTT CONNECT packet by adding the size of
+             * the "Remaining Length" field plus 1 byte for the "Packet Type" field. */
+            connectPacketSize += 1U + remainingLengthEncodedSize( connectPacketSize );
+
+            /**
+             * 268,435,455 - max remaining length  according to spec MQTT-v5;
+             *
+             * */
+            if( remainingLength > MQTT_MAX_REMAINING_LENGTH )
+            {
+                status = MQTTBadParameter;
+            }
+        }
+
+        if( status == MQTTSuccess )
+        {
+            *pRemainingLength = remainingLength;
+            *pPacketSize = connectPacketSize;
+
+            LogDebug( ( "CONNECT packet remaining length=%lu and packet size=%lu.",
+                        ( unsigned long ) *pRemainingLength,
+                        ( unsigned long ) *pPacketSize ) );
+        }
+
+        return status;
+    }
+
+    uint8_t * MQTT_SerializePublishProperties( const MQTTPublishInfo_t * pPublishInfo,
+                                               uint8_t * pIndex )
+    {
+        uint8_t * pIndexLocal = pIndex;
```

Comment:
How are we ensuring pIndexLocal or pIndex has enough space to do all the writes below?